cookbook/core/control_flow/dynamics.py

```diff
@@ -1,56 +1,119 @@
+# #. Derive the number of common characters by comparing the two frequency lists
+#
+# The looping is dependent on the number of characters of both the strings which aren't known until the run time. If the ``@task`` decorator is used to encapsulate the calls mentioned above, the compilation will fail very early on due to the absence of the literal values.
+# Therefore, ``@dynamic`` decorator has to be used.
+#
+# Dynamic workflow is effectively both a task and a workflow. The key thing to note is that the ``body of tasks is run at run time and the
+# body of workflows is run at compile (aka registration) time``. Essentially, this is what a dynamic workflow leverages -- it’s a workflow that is compiled at run time (the best of both worlds)!
 #
-# Local execution will also work as it's treated like a task that will be run with Python native inputs.
+# At execution (run) time, flytekit runs the compilation step, and produces
+# a ``WorkflowTemplate`` (from the dynamic workflow), which flytekit then passes back to Flyte Propeller for further running, much like how sub-workflows are handled.
 #
-# The name of this function was intentionally chosen. Even though these are implemented as tasks in flytekit, you
-# should think about them as workflows. At execution time, flytekit is running the compilation step, and producing
-# a ``WorkflowTemplate``, which it then passes back to Flyte Propeller for further running, much like how sub-workflows
-# are handled.
+# We now define the dynamic workflow encapsulating the above mentioned points.
 @dynamic
-def my_subwf(a: int) -> (typing.List[str], int):
-    s = []
-    for i in range(a):
-        s.append(t1(a=i))
-    return s, 5
+def count_characters(s1: str, s2: str) -> int:
+    """
+    Calls the required tasks and returns the final result"""
 
+    # s1 and s2 are accessible
 
+    # initiliazing an empty list consisting of 26 empty slots corresponding to every alphabet (lower and upper case)
+    freq1 = [0] * 26
+    freq2 = [0] * 26
+
+    # looping through the string s1
+    for i in range(len(s1)):
+
+        # index and freq1 are not accesible as they are promises
+        index = return_index(character=s1[i])
+        freq1 = prepare_list(freq_list=freq1, list_index=index)
+
+    # looping through the string s2
+    for i in range(len(s2)):
+
+        # index and freq2 are not accesible as they are promises
+        index = return_index(character=s2[i])
+        freq2 = prepare_list(freq_list=freq2, list_index=index)
+
+    # counting the common characters
+    return derive_count(freq1=freq1, freq2=freq2)
+
+
+# %%
+# When tasks are called within any workflow, they return Promise objects. Likewise, in a dynamic workflow, the tasks' outputs are Promise objects that cannot be directly accessed (they shall be fulfilled by Flyte later). To get around this problem, the values need to be passed to the other tasks to unwrap them.
```

Comment:
> Promise objects that cannot be directly accessed

to really hit the point home, can we add an example of an operation that isn't valid because of this fact? For example:

_... Promise objects that cannot be directly accessed (they shall be fulfilled by Flyte later). Because of this fact, operations on the `index` variable like `index + 1` are not valid. To get around this problem..._

> A point to note is, local execution will work when a ``@dynamic`` decorator is used because Flytekit treats it like a task that shall run with the Python native inputs

I have a question about this: does this mean that `index + 1` would work locally but not remotely?

Reply:
1. Added the example.
2. `index + 1` works locally because `@dynamic` is treated as `@task`. Refer to lines 107 and 108.